x-pack/plugins/alerting/server/rules_client/methods/bulk_edit.ts

```diff
@@ -302,209 +306,419 @@ async function bulkEditOcc<Params extends RuleTypeParams>(
     );
 
   const rules: Array<SavedObjectsBulkUpdateObject<RawRule>> = [];
+  const skipped: BulkActionSkipResult[] = [];
   const errors: BulkOperationError[] = [];
-  const apiKeysToInvalidate: string[] = [];
-  const apiKeysMap = new Map<string, { oldApiKey?: string; newApiKey?: string }>();
+  const apiKeysMap: ApiKeysMap = new Map();
   const username = await context.getUserName();
```

Comment:
I think you can create this variable inside `updateRuleAttributesAndParamsInMemory`. We do not need to pass it as a parameter. We already pass `context` to `updateRuleAttributesAndParamsInMemory`

Reply:
Not sure I understand you but I don't think I can: I need to acumualte the apiKeys of all the rules we loop through. `apiKeysMap` then gets passed down to `saveBulkUpdatedRules` to create `apiKeysToInvalidate`

Reply:
I am talking about username. We do not use it here, but use inside `updateRuleAttributesAndParamsInMemory`. No need to pass it there. We can just call `context.getUserName` inside `updateRuleAttributesAndParamsInMemory`

Reply:
Ok, now I understood. The reason I'm placing it outside `updateRuleAttributesAndParamsInMemory` is because the username will be the same for all rules being edited. If I call `context.getUserName` inside `updateRuleAttributesAndParamsInMemory` it is an additional async call that will be called for each rule. If we are editing hundreds of rules, it will sum up. I'd rather call it once and pass the calculated value to the function.